c/tskit/trees.c

```diff
@@ -8145,3 +8165,339 @@ tsk_treeseq_extend_edges(const tsk_treeseq_t *self, int max_iter,
+    int *inside = NULL;
+    int *outside = NULL;
+    double *coalescing_pairs = NULL;
+    double *nodes_weight = NULL;
+
+    /* check inputs */
+    ret = tsk_treeseq_check_windows(self, num_windows, windows, options);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = check_time_windows(num_time_windows, time_windows);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = check_set_indexes(num_sample_sets, 2 * num_set_indexes, set_indexes);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_treeseq_check_sample_sets(
+        self, num_sample_sets, sample_set_sizes, sample_sets);
+    if (ret != 0) {
+        goto out;
+    }
+    /* TODO: set defaults? */
+
+    /* map nodes to sample sets */
+    nodes_sample_set = tsk_malloc(num_nodes * sizeof(*nodes_sample_set));
+    if (nodes_sample_set == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+    ret = get_sample_set_index_map(self, num_sample_sets, sample_set_sizes, sample_sets,
+        &total_samples, nodes_sample_set);
+    if (ret != 0) {
+        goto out;
+    }
+
+    /* initialize internal state */
+    inp = (tsk_id_t) num_sample_sets;
+    col = (tsk_id_t) num_set_indexes;
+    nodes_parent = tsk_malloc(num_nodes * sizeof(*nodes_parent));
+    nodes_sample = tsk_malloc(num_nodes * (tsk_size_t) inp * sizeof(*nodes_sample));
+    sample_count = tsk_malloc(num_nodes * (tsk_size_t) inp * sizeof(*sample_count));
+    if (nodes_parent == NULL || nodes_sample == NULL || sample_count == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+    tsk_memset(nodes_sample, 0, num_nodes * (tsk_size_t) inp * sizeof(*nodes_sample));
+    for (i = 0; i != (tsk_id_t) num_nodes; i++) {
+        if (nodes_sample_set[i] != TSK_NULL) {
+            nodes_sample[i * inp + nodes_sample_set[i]] = 1;
+        }
+    }
+    tsk_memcpy(sample_count, nodes_sample,
+        num_nodes * (tsk_size_t) inp * sizeof(*sample_count));
+    for (i = 0; i != (tsk_id_t) num_nodes; i++) {
+        nodes_parent[i] = TSK_NULL;
+    }
+
+    /* map nodes to time windows */
+    nodes_time_window = tsk_malloc(num_nodes * sizeof(*nodes_time_window));
+    if (nodes_time_window == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+    ret = get_time_windows_index_map(
+        self, &num_time_windows, time_windows, nodes_time_window);
+    if (ret != 0) {
+        goto out;
+    }
+
+    /* initialize outputs */
+    row = (tsk_id_t) num_time_windows;
+    inside = tsk_malloc((tsk_size_t) col * sizeof(inside));
+    outside = tsk_malloc((tsk_size_t) col * sizeof(outside));
+    if (inside == NULL || outside == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+    coalescing_pairs
+        = tsk_malloc((tsk_size_t) row * (tsk_size_t) col * sizeof(coalescing_pairs));
+    nodes_weight
+        = tsk_malloc((tsk_size_t) row * (tsk_size_t) col * sizeof(nodes_weight));
+    if (coalescing_pairs == NULL || nodes_weight == NULL) {
```

Comment:
I think you can merge together some of these malloc/check things. I think it's fine to malloc up to (say) 10 things at once and check them all for NULL just after. We need to review these manually anyway, and it saves some noise in the coverage reports.